text/0192-remove-constructs-compat.md

```diff
@@ -0,0 +1,667 @@
+returns the list of error messages (whether or not the construct implements
+`IValidation`).
+
+## Summary
+
+The following table summarizes the API changes between 1.x and 2.x:
+
+1.x|2.x
+------|-----
+`@aws-cdk/*` | `aws-cdk-lib` and `constructs@^4`
+`import { Construct } from '@aws-cdk/core'` | `import { Construct } from 'constructs'`
+`@aws-cdk/core.Construct` | `constructs.Construct`
+`@aws-cdk/core.IConstruct` | `constructs.IConstruct`
+`@aws-cdk/core.ConstructOrder` | `constructs.ConstructOrder`
+`@aws-cdk/core.ConstructNode` | `constructs.Node`
+`Construct.isConstruct(x)` | `x instanceof Construct`
+`construct.node.applyAspect(aspect)` | `Aspects.of(construct).apply(aspect)`
+`@aws-cdk/core.IDependable` | `constructs.IDependable`
+`@aws-cdk/core.DependencyTrait` | `constructs.Dependable`
+`@aws-cdk.core.DependencyTrait.get(x)` | `constructs.Dependable.of(x)`
+`construct.node.dependencies` | Is now non-transitive
+`construct.addMetadata()` | Stack trace not attached by default
+`ConstructNode.prepare(node)`, `onPrepare()`, `prepare()` | Not supported
+`ConstructNode.synthesize(node)`, `onSynthesize()`, `synthesize()` | Not supported
+`construct.onValidate()`, `construct.validate()` hooks | Implement `constructs.IValidation`
+`ConstructNode.validate(node)` | `construct.node.validate()`
+
+# Motivation
+
+There are various motivations for this change:
+
+1. Removal of a redundant layer
+2. User confusion caused by multiple `Construct` types
+3. Inability to compose AWS CDK constructs into other domains
+4. Non-intuitive dependency requirement on `constructs`
+
+### 1. Redundant layer
+
+The current compatibility layer does not have any logic in it. It is pure glue
+introduced in order to avoid breaking v1.x users. As we release v2.0 we are able
+to clean up this layer, and with it, improve maintainability and code hygiene.
+
+### 2. Multiple `Construct` types
+
+The current situation is error-prone since we have two `Construct` classes in
+the type closure. For example, when a developer types `"Construct"` and uses
+VSCode to _automatically add import statements_, the IDE will actually add an
+import for `constructs.Construct`, which is _not_ the type they need. Using the
+wrong type will fail during compilation (similar error as above).
+
+### 3. Composability with other domains
+
+The main motivation for this change is to enable composition of AWS CDK
+constructs with constructs from other domains.
+
+It is currently impossible to define AWS CDK constructs within a non-AWS-CDK
+construct scope.
+
+For example, consider the
+[Terrastack](https://github.com/TerraStackIO/terrastack) project or a similar
+one, which uses CDK constructs to define stacks through Terraform. Say we want to
+use an AWS CDK L2 inside a Terraform stack construct:
+
+```ts
+const stack = new terrastack.Stack(...);
+
+// COMPILATION ERROR: `this` is not a `cdk.Construct`.
+new s3.Bucket(this, 'my-bucket');
```

Comment:
> Have we considered properly defining the constructs.IConstruct interface

What do you mean by properly defining it?

> using that as the type of the scope argument

I have considered it but I think it will increase the cognitive load in defining custom constructs (which is very important to keep low) and is also potentially error prone (i.e. if I *didn't* define the interface, it would still work but it will limit composability).

I rather keep the current model in which implementing a custom construct is just `extend Construct` + `scope: Construct`.

> avoid having to take a peerDependency on constructs?

The peer dependency is required because `construct.Construct` is used as a base class and therefore it's public API is by definition part of the public API of any derived classes.

Reply:
👍